crates/holochain_sqlite/src/db.rs

```diff
@@ -201,43 +207,109 @@ pub trait WriteManager<'e> {
     #[cfg(feature = "test_utils")]
     fn with_commit_test<R, F>(&'e mut self, f: F) -> Result<R, DatabaseError>
     where
-        F: 'e + FnOnce(&mut Transaction) -> R,
+        F: 'e + Clone + FnOnce(&mut Transaction) -> R,
     {
         self.with_commit(|w| DatabaseResult::Ok(f(w)))
     }
 }
 
-impl<'e> ReadManager<'e> for PConn {
+impl<'e> ReadManager<'e> for DbRead {
     fn with_reader<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
         E: From<DatabaseError>,
         F: 'e + FnOnce(Transaction) -> Result<R, E>,
     {
-        let txn = self.transaction().map_err(DatabaseError::from)?;
-        f(txn)
+        Ok(self.conn()?.with_reader(f)?)
     }
+}
 
-    #[cfg(feature = "test_utils")]
-    fn with_reader_test<R, F>(&'e mut self, f: F) -> R
+impl<'e> WriteManager<'e> for DbWrite {
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
-        F: 'e + FnOnce(Transaction) -> R,
+        E: From<DatabaseError> + std::fmt::Debug,
+        F: 'e + Clone + FnOnce(&mut Transaction) -> Result<R, E>,
     {
-        self.with_reader(|r| DatabaseResult::Ok(f(r))).unwrap()
+        Ok(self.conn()?.with_commit(f)?)
     }
 }
 
-impl<'e> WriteManager<'e> for PConn {
-    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
+impl<'e> ReadManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_reader<E, R, F>(&'e mut self, f: F) -> Result<R, E>
     where
         E: From<DatabaseError>,
-        F: 'e + FnOnce(&mut Transaction) -> Result<R, E>,
+        F: 'e + FnOnce(Transaction) -> Result<R, E>,
     {
-        let mut txn = self
-            .transaction_with_behavior(TransactionBehavior::Exclusive)
+        tracing::trace!("Attempting to acquire read lock on TXN_MUTEX.");
+        let _g = TXN_MUTEX.read();
+        tracing::trace!("Got read lock on TXN_MUTEX.");
+        let txn = self
+            .transaction_with_behavior(TransactionBehavior::Deferred)
             .map_err(DatabaseError::from)?;
-        let result = f(&mut txn)?;
-        txn.commit().map_err(DatabaseError::from)?;
-        Ok(result)
+        let result = f(txn);
+        tracing::trace!("Released read lock on TXN_MUTEX.");
+        result
+    }
+}
+
+impl<'e> WriteManager<'e> for PConn {
+    #[tracing::instrument(skip(self, f))]
+    fn with_commit<E, R, F>(&'e mut self, f: F) -> Result<R, E>
```

Comment:
I know we've been kinda deferring dealing with all the synchronous sql code vs tokio, but both the RwLocks and the backoff retry add additional layers that should be in spawn_blocking blocks and `with_reader` and `with_commit` made `async`.

Reply:
Sounds like we decided to asyncify things in a followup PR